[processor/dnslookup] Add DNS resolver with hostfile, nameserver, and caching support

kaisecheng opened the pull request:
#### Description

This PR introduces the core DNS resolution functionality for use in the `dnslookupprocessor`.
- Resolver interface: Defines forward and reverse DNS lookup operations.
- Hostfile resolver: Supports resolving from one or more hostfiles.
   When hostname is not in files, the processor continues to the next resolver in the chain. 
- Nameserver resolver: Supports custom DNS servers and system resolver.
   It performs A/CNAME record lookups to resolve hostnames to IP addresses, 
      and PTR record lookups for reverse DNS resolution (IP to hostname)
   It retries with exponential backoff on failures.
   The resolver returns all addresses including IPv4 and IPv6.
- Chain resolver: Attempts resolution through a sequence of resolvers.
   Designed to support a resolution chain such as: hostfile -> nameserver -> system default.
   When the chain gets a valid result, it returns. No resolution is considered a valid result.
   The actual resolver chain initialization logic will be added in the next PR.
- Cache resolver: Cache results from chain resolver.
   Successful resolutions are stored in the hit cache. 
   no resolution / permanent error (DNS config error) are stored in miss cache
   temporary network errors (timeout) are not cached
   Implements LRU caching with TTL support.

The current processor is a no-op, as user configuration and signal processing are not yet implemented

#### Link to tracking issue

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34398

#### Testing

Added unit tests for all resolvers

#### Documentation

kaisecheng committed:
[processor/dnslookup] Implement DNS resolver with hostfile, nameserver, and caching support

This PR introduces the core DNS resolution functionality for use in the `dnslookupprocessor`.
- Resolver interface: Defines forward and reverse DNS lookup operations.
- Hostfile resolver: Supports resolving from one or more hostfiles.
- Nameserver resolver: Supports custom DNS servers and system resolver.
   It retries with exponential backoff on failures.
- Chain resolver: Attempts resolution through a sequence of resolvers.
   Returns the first successful result. No resolution is considered a success
- Cache resolver - Cache results from chain resolvers.
   Successful resolutions are stored in the hit cache. Failed lookups are stored in miss cache.
   Implements LRU caching with TTL support.

kaisecheng committed:
changelog

kaisecheng committed:
make goporto

kaisecheng committed:
make gotidy

edmocosta reviewed (commented):
Thanks for working on that @kaisecheng, it looks great! I've left a few suggestions and questions, and I'm sorry (beforehand) if a few of them does not make much sense, or if I forgot to delete it after answering my own questions :)

kaisecheng commented on processor/dnslookupprocessor/internal/resolver/resolver.go:
```diff
@@ -0,0 +1,44 @@
+	// Returns hostname as string or error if resolution fails
+	Reverse(ctx context.Context, ip string) (string, error)
+
+	Name() string
```
Yes. It mostly for log. System resolver and nameserver resolver use the same class `NameserverResolver`, so `%T` isn't ideal. But as discussed, I will remove most of the debug log, hence, will drop it from interface

kaisecheng commented on processor/dnslookupprocessor/internal/resolver/cache.go:
```diff
@@ -0,0 +1,154 @@
+	// Call the underlying chain resolver
+	result, err := resolveFunc(ctx, target)
+	// Add failure to miss cache
+	if err != nil {
```
Thanks for the suggestion. You are right that not all errors should be in miss cache. I think it makes sense to NOT cache the retryable error like network timeout. 

MIss cache should cache the result of non temporary error and NXDOMAIN non-existent domain

kaisecheng commented on processor/dnslookupprocessor/internal/resolver/chain.go:
```diff
@@ -0,0 +1,87 @@
+	}
+
+	// When the host file resolver is the only one in the chain, and it returns ErrNotInHostFiles,
+	if errors.Is(lastErr, ErrNotInHostFiles) {
```
`ErrNoResolution` is a valid result from NS and indicates that no need to go further for another lookup.

Considering the chain resolver [ hosts, nameservers, system ], throwing ErrNoResolution from nameservers would stop the chain, however, we don't want to stop if an entry is not found in hostfile. In order to distinguish these two cases, ErrNotInHostFiles is used to signal the chain to continue.

I will move this error handling to cache resolver.

edmocosta replied:
> I will move this error handling to cache resolver.

I think that's the same as keeping it here. My comment motivation was based on the chain resolver, which seems to be made to be generic, and not aware of implementation details, such as the `ErrNotInHostFiles`, neither that it must stop if the first resolver return `ErrNoResolution` (as we don't know their types details). 

IMO, this is somehow making things harder to understand, if we only have 3 types of resolvers, being cache a wrapper, maybe we don't need this chain to be generic, and support a list of unknown resolvers at the `NewChainResolver`. If we do need it to be generic, so I think we should be consistent generic.

kaisecheng commented on processor/dnslookupprocessor/internal/resolver/chain.go:
```diff
@@ -0,0 +1,87 @@
+		result, err := resolverFn(r)
+
+		// Successful resolution
+		if err == nil || errors.Is(err, ErrNoResolution) {
```
As discussed, `ErrNoResolution` is considered as a valid result that no need to retry.

kaisecheng commented on processor/dnslookupprocessor/internal/resolver/resolver.go:
```diff
@@ -0,0 +1,44 @@
+type Resolver interface {
+	// Resolve performs forward DNS resolution (hostname to IP)
+	// Returns IP address as string or error if resolution fails
+	Resolve(ctx context.Context, hostname string) (string, error)
```
Yes, users could pick the one base on their need. This requires change on hostfile handling. Maybe we could improve it later?

edmocosta commented on processor/dnslookupprocessor/internal/resolver/nameserver.go:
```diff
@@ -0,0 +1,277 @@
+		}
+
+		// Find first IPv4 address if available
+		for _, ip := range ips {
```
The issues isn't the parsing operation, it's the string representation of those IPv6 (`.String()`). I didn't know, but it seems the standard Go's library does fully comply with the RFC 5952, which is okay I guess, but makes IPv6 like  `0:0:0:0:0000:ffff:127.1.2.3` to be formatted as IPv4 `127.1.2.3` by default, instead of `::ffff:127.1.2.3`. 

That said, I don't think we should strictly follow the RFC as well, so the standard output should be fine here. But I'm still concerned about prioritizing IPv4 over IPv6, and not offering users any config to setup that.

kaisecheng commented on processor/dnslookupprocessor/internal/resolver/nameserver.go:
```diff
@@ -0,0 +1,277 @@
+func (r *NameserverResolver) Reverse(ctx context.Context, ip string) (string, error) {
+	return r.lookupWithNameservers(ctx, ip, LogKeyIP, func(resolver Lookup, fnCtx context.Context) (string, error) {
+		hostnames, err := resolver.LookupAddr(fnCtx, ip)
+		if err != nil && hostnames == nil {
```
`net.LookupAddr()` filtered out invalid domain names and returns error alongside with valid results. `len(hostnames) > 0` is sure for at least one valid result

edmocosta replied:
Oh! I didn't know that, thank you!

kaisecheng committed:
Enhanced error handling in nameserver and cache resolver

Handled non retryable errors in nameserver resolver

Cache resolver
  - Moved NXDOMAIN/ErrNoResolution from hit cache to miss cache
  - Temporary network failures are no longer cached
  - Permanent error, e.g. DNS config error stored in miss cache

kaisecheng committed:
Add support to return multiple resolutions

- Changed the return type of the Resolver interface to []string
- Resolve() returns all resolutions from nameserver rather than prioritizing IPv4 or picking the first one

edmocosta commented on processor/dnslookupprocessor/internal/resolver/nameserver.go:
```diff
@@ -0,0 +1,277 @@
+
+	resolvers := make([]Lookup, len(normalizeNameservers))
+	for i, ns := range normalizeNameservers {
+		nameserver := ns // copy to local
```
The value is already being copied on the `newNetResolver` call, and considering it's running Go 1.23, even if we captured it here, it shouldn't be an issue anymore (loopvar).

kaisecheng committed:
Remove debug log

- Remove Name() from Resolver interface
- Remove log specific params and helper

kaisecheng committed:
Remove Close() from Resolver interface

`go-freelru` does not require purge() to free all resources.
The lifecycle of exporter can manage cleaning up everything.

kaisecheng committed:
Add error log for non-temporary network issue

kaisecheng committed:
Decoupled cache and chain from implementations details of resolvers

Moved error handling to hostfile and nameserver.
Resolve() and Reverse() return
  - ([]string, nil) on successful resolution
  - (nil, error) if an error occurred
  - (nil, ErrNoResolution) if server replied with no resolution
  - (nil, nil) if no resolution is found and expected the caller to try the next resolver

kaisecheng committed:
Replaced ErrNoResolution with empty slice

nameserver returns ([]string{}, nil) to indicate the server responded, but returned no resolution.
It is a valid answer from DNS. The chain should stop and return the answer.

Resolve() and Reverse() return
  - ([]string, nil) on successful resolution
  - ([]string{}, nil) if the server responded, but returned no resolution
  - (nil, nil) if no resolution is found and expected the caller to try the next resolver
  - (nil, error) if an error occurred

edmocosta reviewed (commented):
LGTM, I've ran a few local tests and it seems to be working as expected. Great work! 
I think we just need to wait for @andrzej-stencel review and we're ready to go. Thank you!

kaisecheng committed:
Add unit tests for Reverse()

Resolve() has extensive tests covering the share part of Reverse().
These additional tests are to ensure Reverse() has been called.

andrzej-stencel reviewed (changes requested):
This PR apparently adds a bunch of code to the component that is not ever used. I'm guessing this is in an effort to make incremental steps instead of one big PR with tons of code. However, I don't think this is the right way to make incremental steps in the implementation of the component. As much as possible, each PR should add code that can delivers a user-visible change in behavior of the component. Even the first skeleton PR delivers a change: it allows users to mention a component by its name in the config. The next PR after the skeleton PR should deliver the next minimum viable functional change: in the case of the DNS Lookup processor, this might be one resolver, perhaps even a dummy one, and it should be possible to actually use it.

The DNS Lookup processor currently does not have a README file and this PR does not add one either. Adding a README should be one of the first steps when creating a new component.

How does this PR relate to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39642?

kaisecheng commented:
Thanks for the guideline Andrzej. I will close this and create a smaller PR with basic processor and one resolver.

> How does this PR relate to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39642?

That draft includes the implementation of both the processor and the resolvers. My plan was to split it into two smaller PRs. This PR, focused on the resolvers, is one of them.
